File: services/access/src/personalization/schemas.py
from __future__ import annotations

import re
from enum import Enum
from typing import Annotated, Any
from uuid import UUID
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from ninja import Schema
from pydantic import Field, field_validator


class ThemeEnum(str, Enum):
    """Available theme options."""
    LIGHT = "light"
    DARK = "dark"
    AUTO = "auto"


class ThemeSourceEnum(str, Enum):
    """Available theme source options."""
    PORTAL = "portal"
    ID = "id"


class LanguageEnum(str, Enum):
    """Supported interface languages."""
    EN = "en"
    RU = "ru"


class FontSizeEnum(str, Enum):
    """Available font size options."""
    SMALL = "small"
    MEDIUM = "medium"
    LARGE = "large"


class ProfileVisibilityEnum(str, Enum):
    """Profile visibility options."""
    PUBLIC = "public"
    MEMBERS = "members"
    PRIVATE = "private"


class EmailDigestEnum(str, Enum):
    """Email digest frequency options."""
    INSTANT = "instant"
    HOURLY = "hourly"
    DAILY = "daily"
    WEEKLY = "weekly"


class NotificationChannelEnum(str, Enum):
    """Notification delivery channels."""
    EMAIL = "email"
    IN_APP = "in_app"
    PUSH = "push"


# Regex for hex color validation
HEX_COLOR_PATTERN = re.compile(r"^#[0-9A-Fa-f]{6}$")

# Common IANA timezone names for quick validation
COMMON_TIMEZONES = frozenset([
    "UTC", "GMT",
    "Europe/Moscow", "Europe/London", "Europe/Paris", "Europe/Berlin",
    "Europe/Tallinn", "Europe/Helsinki", "Europe/Riga", "Europe/Vilnius",
    "America/New_York", "America/Chicago", "America/Denver", "America/Los_Angeles",
    "America/Toronto", "America/Vancouver", "America/Sao_Paulo",
    "Asia/Tokyo", "Asia/Seoul", "Asia/Shanghai", "Asia/Singapore",
    "Asia/Dubai", "Asia/Kolkata", "Asia/Hong_Kong",
    "Australia/Sydney", "Australia/Melbourne", "Australia/Perth",
    "Pacific/Auckland", "Pacific/Honolulu",
    "Africa/Johannesburg", "Africa/Cairo",
])


def validate_timezone(tz: str) -> str:
    """Validate IANA timezone string."""
    if tz in COMMON_TIMEZONES:
        return tz
    try:
        ZoneInfo(tz)
        return tz
    except ZoneInfoNotFoundError:
        raise ValueError(f"Invalid timezone: {tz}")


def validate_hex_color(color: str) -> str:
    """Validate hex color format (#RRGGBB)."""
    if not HEX_COLOR_PATTERN.match(color):
        raise ValueError(f"Invalid hex color format: {color}. Expected #RRGGBB")
    return color.upper()


class NotificationChannelConfigSchema(Schema):
    """Configuration for a notification channel."""
    enabled: bool = True
    channels: list[NotificationChannelEnum] = Field(default_factory=list)

    @field_validator("channels", mode="before")
    @classmethod
    def validate_channels(cls, v: list[str]) -> list[str]:
        valid_channels = {c.value for c in NotificationChannelEnum}
        for channel in v:
            if channel not in valid_channels:
                raise ValueError(f"Invalid channel: {channel}. Valid: {valid_channels}")
        return v


class EmailChannelSchema(Schema):
    """Email notification channel settings."""
    enabled: bool = True
    digest: EmailDigestEnum = EmailDigestEnum.INSTANT


class InAppChannelSchema(Schema):
    """In-app notification channel settings."""
    enabled: bool = True


class PushChannelSchema(Schema):
    """Push notification channel settings (MVP: disabled)."""
    enabled: bool = False


class QuietHoursSchema(Schema):
    """Quiet hours configuration for notifications."""
    enabled: bool = False
    start: str = "22:00"  # HH:MM format
    end: str = "08:00"

    @field_validator("start", "end", mode="before")
    @classmethod
    def validate_time_format(cls, v: str) -> str:
        if not re.match(r"^([01]\d|2[0-3]):[0-5]\d$", v):
            raise ValueError(f"Invalid time format: {v}. Expected HH:MM (24h)")
        return v


class NotificationSettingsSchema(Schema):
    """Full notification settings structure."""
    email: EmailChannelSchema = Field(default_factory=EmailChannelSchema)
    in_app: InAppChannelSchema = Field(default_factory=InAppChannelSchema)
    push: PushChannelSchema = Field(default_factory=PushChannelSchema)
    types: dict[str, Any] = Field(default_factory=dict)
    quiet_hours: QuietHoursSchema = Field(default_factory=QuietHoursSchema)


class AppearanceSchema(Schema):
    """User appearance/theme preferences."""
    theme: ThemeEnum = ThemeEnum.AUTO
    theme_source: ThemeSourceEnum = ThemeSourceEnum.PORTAL
    accent_color: Annotated[str, Field(pattern=r"^#[0-9A-Fa-f]{6}$")] = "#8B5CF6"
    font_size: FontSizeEnum = FontSizeEnum.MEDIUM
    high_contrast: bool = False
    reduce_motion: bool = False

    @field_validator("accent_color", mode="before")
    @classmethod
    def normalize_color(cls, v: str) -> str:
        return validate_hex_color(v)


class LocalizationSchema(Schema):
    """User localization preferences."""
    language: LanguageEnum = LanguageEnum.EN
    timezone: str = "UTC"

    @field_validator("timezone", mode="before")
    @classmethod
    def validate_tz(cls, v: str) -> str:
        return validate_timezone(v)


class PrivacySchema(Schema):
    """User privacy preferences."""
    profile_visibility: ProfileVisibilityEnum = ProfileVisibilityEnum.MEMBERS
    show_online_status: bool = True
    show_vote_history: bool = False
    share_activity: bool = True
    allow_mentions: bool = True
    analytics_enabled: bool = True
    recommendations_enabled: bool = True


class UserPreferenceSchema(Schema):
    """Full user preference response schema."""

    id: UUID
    user_id: UUID
    tenant_id: UUID
    appearance: AppearanceSchema
    localization: LocalizationSchema
    notifications: dict[str, Any]
    privacy: PrivacySchema
    created_at: str
    updated_at: str


class UserPreferenceUpdateSchema(Schema):
    """Schema for updating user preferences (partial updates supported)."""

    appearance: AppearanceSchema | None = None
    localization: LocalizationSchema | None = None
    notifications: dict[str, Any] | None = None
    privacy: PrivacySchema | None = None


class UserPreferenceDefaultsSchema(Schema):
    """Schema for default preferences response."""

    appearance: AppearanceSchema
    localization: LocalizationSchema
    notifications: NotificationSettingsSchema
    privacy: PrivacySchema

File: web/portal-frontend/src/features/personalization/validation/schemas.ts
/**
 * Zod schemas for user preferences validation
 */
import { z } from 'zod';

// Base schemas for individual types
export const themeSchema = z.enum(['light', 'dark', 'auto']);
export const themeSourceSchema = z.enum(['portal', 'id']);
export const languageSchema = z.enum(['en', 'ru']);
export const fontSizeSchema = z.enum(['small', 'medium', 'large']);
export const profileVisibilitySchema = z.enum(['public', 'members', 'private']);
export const emailDigestSchema = z.enum(['instant', 'hourly', 'daily', 'weekly']);
export const notificationChannelSchema = z.enum(['email', 'in_app', 'push']);

// Color validation (hex format)
export const hexColorSchema = z.string().regex(/^#[0-9A-Fa-f]{6}$/, 'Invalid hex color format');

// Timezone validation (IANA timezone names)
export const timezoneSchema = z.string().min(1, 'Timezone is required').refine(
  (timezone) => {
    try {
      Intl.DateTimeFormat(undefined, { timeZone: timezone });
      return true;
    } catch {
      return false;
    }
  },
  'Invalid timezone'
);

// Time format validation (HH:MM)
export const timeFormatSchema = z.string().regex(/^([01]?[0-9]|2[0-3]):[0-5][0-9]$/, 'Invalid time format (use HH:MM)');

// Appearance settings schema
export const appearanceSettingsSchema = z.object({
  theme: themeSchema.optional(),
  theme_source: themeSourceSchema.optional(),
  accent_color: hexColorSchema.optional(),
  font_size: fontSizeSchema.optional(),
  high_contrast: z.boolean().optional(),
  reduce_motion: z.boolean().optional(),
});

// Localization settings schema  
export const localizationSettingsSchema = z.object({
  language: languageSchema.optional(),
  timezone: timezoneSchema.optional(),
});

// Notification channel config schema
export const notificationChannelConfigSchema = z.object({
  enabled: z.boolean(),
  channels: z.array(notificationChannelSchema).min(0),
});

// Email notification settings schema
export const emailNotificationSettingsSchema = z.object({
  enabled: z.boolean().optional(),
  digest: emailDigestSchema.optional(),
});

// In-app notification settings schema
export const inAppNotificationSettingsSchema = z.object({
  enabled: z.boolean().optional(),
});

// Push notification settings schema
export const pushNotificationSettingsSchema = z.object({
  enabled: z.boolean().optional(),
});

// Quiet hours settings schema
export const quietHoursSettingsSchema = z.object({
  enabled: z.boolean().optional(),
  start: timeFormatSchema.optional(),
  end: timeFormatSchema.optional(),
}).refine(
  (data) => {
    if (!data.enabled) return true;
    if (!data.start || !data.end) return false;
    
    // Validate that start and end times are different
    return data.start !== data.end;
  },
  {
    message: 'Start and end times must be different',
    path: ['end'],
  }
);

// Notification type config schema (for polls, events, etc.)
export const notificationTypeConfigSchema = z.record(z.string(), notificationChannelConfigSchema);

// Full notification settings schema
export const notificationSettingsSchema = z.object({
  email: emailNotificationSettingsSchema.optional(),
  in_app: inAppNotificationSettingsSchema.optional(),
  push: pushNotificationSettingsSchema.optional(),
  types: z.object({
    polls: notificationTypeConfigSchema.optional(),
    events: notificationTypeConfigSchema.optional(),
    community: notificationTypeConfigSchema.optional(),
    system: notificationTypeConfigSchema.optional(),
  }).optional(),
  quiet_hours: quietHoursSettingsSchema.optional(),
});

// Privacy settings schema
export const privacySettingsSchema = z.object({
  profile_visibility: profileVisibilitySchema.optional(),
  show_online_status: z.boolean().optional(),
  show_vote_history: z.boolean().optional(),
  share_activity: z.boolean().optional(),
  allow_mentions: z.boolean().optional(),
  analytics_enabled: z.boolean().optional(),
  recommendations_enabled: z.boolean().optional(),
});

// Main preferences update schema
export const preferencesUpdateSchema = z.object({
  appearance: appearanceSettingsSchema.optional(),
  localization: localizationSettingsSchema.optional(),
  notifications: notificationSettingsSchema.optional(),
  privacy: privacySettingsSchema.optional(),
});

// Full user preferences schema (for validation of complete data)
export const userPreferencesSchema = z.object({
  id: z.string().uuid(),
  user_id: z.string().uuid(),
  tenant_id: z.string().uuid(),
  appearance: z.object({
    theme: themeSchema,
    theme_source: themeSourceSchema,
    accent_color: hexColorSchema,
    font_size: fontSizeSchema,
    high_contrast: z.boolean(),
    reduce_motion: z.boolean(),
  }),
  localization: z.object({
    language: languageSchema,
    timezone: timezoneSchema,
  }),
  notifications: z.object({
    email: z.object({
      enabled: z.boolean(),
      digest: emailDigestSchema,
    }),
    in_app: z.object({
      enabled: z.boolean(),
    }),
    push: z.object({
      enabled: z.boolean(),
    }),
    types: z.object({
      polls: notificationTypeConfigSchema.optional(),
      events: notificationTypeConfigSchema.optional(),
      community: notificationTypeConfigSchema.optional(),
      system: notificationTypeConfigSchema.optional(),
    }),
    quiet_hours: z.object({
      enabled: z.boolean(),
      start: timeFormatSchema,
      end: timeFormatSchema,
    }),
  }),
  privacy: z.object({
    profile_visibility: profileVisibilitySchema,
    show_online_status: z.boolean(),
    show_vote_history: z.boolean(),
    share_activity: z.boolean(),
    allow_mentions: z.boolean(),
    analytics_enabled: z.boolean(),
    recommendations_enabled: z.boolean(),
  }),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

// Inferred TypeScript types from schemas
export type AppearanceSettingsInput = z.infer<typeof appearanceSettingsSchema>;
export type LocalizationSettingsInput = z.infer<typeof localizationSettingsSchema>;
export type NotificationSettingsInput = z.infer<typeof notificationSettingsSchema>;
export type PrivacySettingsInput = z.infer<typeof privacySettingsSchema>;
export type PreferencesUpdateInput = z.infer<typeof preferencesUpdateSchema>;
export type UserPreferencesInput = z.infer<typeof userPreferencesSchema>;
